src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs

```diff
@@ -136,486 +121,350 @@ public bool AllowAutoRedirect
                     {
-                        foreach (KeyValuePair<string, IEnumerable<string>> header in request.Headers)
-                        {
-                            foreach (string value in header.Value)
-                            {
-                                jsHeaders.Invoke("append", header.Key, value);
-                            }
-                        }
-                        if (request.Content != null)
-                        {
-                            foreach (KeyValuePair<string, IEnumerable<string>> header in request.Content.Headers)
-                            {
-                                foreach (string value in header.Value)
-                                {
-                                    jsHeaders.Invoke("append", header.Key, value);
-                                }
-                            }
-                        }
-                        requestObject.SetObjectProperty("headers", jsHeaders);
+                        optionNames.Add(item.Key);
+                        optionValues.Add(item.Value);
                     }
-
-
-                    JSObject abortController = new JSObject("AbortController");
-                    using JSObject signal = (JSObject)abortController.GetObjectProperty("signal");
-                    requestObject.SetObjectProperty("signal", signal);
-
-                    abortRegistration = cancellationToken.Register(() =>
+                }
+                if (request.RequestUri == null)
+                {
+                    throw new ArgumentNullException(nameof(request.RequestUri));
+                }
+                string uri = request.RequestUri.ToString();
+                Task<IJSObject>? promise;
+                cancellationToken.ThrowIfCancellationRequested();
+                if (request.Content != null)
+                {
+                    if (request.Content is StringContent)
                     {
-                        if (!abortController.IsDisposed)
-                        {
-                            abortController.Invoke("abort");
-                            abortController?.Dispose();
-                        }
-                    });
+                        string body = await request.Content.ReadAsStringAsync(cancellationToken)
+                            .ConfigureAwait(continueOnCapturedContext: true);
+                        cancellationToken.ThrowIfCancellationRequested();
 
-                    using var args = new System.Runtime.InteropServices.JavaScript.Array();
-                    if (request.RequestUri != null)
-                    {
-                        args.Push(request.RequestUri.ToString());
-                        args.Push(requestObject);
+                        promise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames.ToArray(), optionValues.ToArray(), abortController, body);
```

Comment:
i can imagine it being useful in the .net 8 or 9 time period to be able to marshal List<T> across the boundary instead of needing to call ToArray on it